include/alp/utils/parser/MatrixFileIterator.hpp

```diff
@@ -0,0 +1,377 @@
+					}
+
+					// otherwise, not equal
+					return false;
+				}
+
+				/** Standard check for inequality, relies on equality check. */
+				bool operator!=( const MatrixFileIterator &x ) const {
+					return ! ( operator==( x ) );
+				}
+
+				// this assumes full triplet data
+				MatrixFileIterator & operator++() {
+					// if ended, stop
+					if( ended ) {
+						return *this;
+					}
+					// if this is the first function call on this iterator, call preprocess first
+					if( started ) {
+						preprocess();
+						started = false;
+						(void)operator++();
+					}
+
+					if( ! ( infile >> val ) ) {
+						ended = true;
+					} else {
+#ifdef _DEBUG
+						T temp = val;
+						converter( temp );
+						std::cout << "MatrixFileIterator::operator++  parsed line ``"
+							  << val  << "'', with value after conversion "
+							  << temp << "\n";
+#endif
+						// convert value
+						converter( val );
+
+						if(
+							properties._symmetry == MatrixFileProperties::MMsymmetries::SYMMETRIC ||
+							properties._symmetry == MatrixFileProperties::MMsymmetries::HERMITIAN
+						) {
+							++rowidx;
+							if( rowidx  == properties._n + 1 ) {
+								++colidx;
+								rowidx = colidx + 1;
+							}
+						} else if (
+							properties._symmetry == MatrixFileProperties::MMsymmetries::SKEWSYMMETRIC
+						) {
+							throw std::runtime_error(
+								"Not implemented i,j: SKEWSYMMETRIC."
+							);
+						} else if (
+							properties._symmetry == MatrixFileProperties::MMsymmetries::GENERAL
+						) {
+							++rowidx;
+							if( rowidx == properties._m + 1 ) {
+								rowidx = 1;
+								++colidx;
+							}
+
+						}
+
+					}
+
+#ifdef _DEBUG
+					std::cout << "MatrixFileIterator::operator++ "
+						": buffer at index "
+						  << i << " now contains " << val << "\n";
+#endif
+
+					// done
+					return *this;
+				}
+
+				/** Standard dereferencing of iterator. */
+				const OutputType & operator*() {
+					if( started ) {
+						preprocess();
+						started = false;
+						(void)operator++();
+					}
+					if( ended ) {
+						throw std::runtime_error(
+							"Attempt to dereference (via operator*) "
+							"MatrixFileIterator in end "
+							"position." );
+					}
+					return val;
+				}
+
+				/** Standard pointer request of iterator. */
+				const OutputType * operator->() {
+					if( started ) {
+						preprocess();
+						started = false;
+						(void)operator++();
+					}
+					if( ended ) {
+						throw std::runtime_error(
+							"Attempt to dereference (via "
+							"operator->) "
+							"MatrixFileIterator in end "
+							"position."
+						);
+					}
+					return &( val );
+				}
+
+				/** Returns the current col index. */
+				const S j() const {
+					if( started ) {
+						const_cast< MatrixFileIterator< T > * >( this )->preprocess();
+						const_cast< MatrixFileIterator< T > * >( this )->started = false;
+						(void)const_cast< MatrixFileIterator< T > * >( this )->operator++();
```

Comment:
Why is `j()` declared as `const` if it modifies the object? I think the standard says that striping the `const` qualifier from a const object may result in an undefined behavior.